### PR TITLE
Support DigitalOcean Block Storage

### PR DESCRIPTION
**Pull request by noahl**
Two major changes together, because they interact:
  - support DigitalOcean Block Storage volumes, with disk type
  'block-storage'
  - switch from raw curl commands to the official DigitalOcean CLI
  client, 'doctl'.

Moving to 'doctl' made it easier to support block-storage volumes, and
also allowed lots of code cleanups.

### CODE REVIEW

**Comment by noahl** [comment]
@ehankland @hildrum @yuyantingzero please take a look.

File: perfkitbenchmarker/providers/digitalocean/digitalocean_disk.py [hunk 1]
```diff
@@ -34,13 +61,71 @@ def Detach(self):
...
+    self.volume_name = 'pkb-%s-%s' % (FLAGS.run_uri, self.disk_number)
+
+    response, retcode = util.DoctlAndParse(
+        ['compute', 'volume', 'create',
+         self.volume_name,
+         '--region', self.zone,
+         '--size', str(self.disk_size) + 'gb'])
+    if retcode:
+      raise errors.Resource.RetryableCreationError(
+          'Error creating disk: %s' % response)
```
Comment (cwilkes) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_disk.py]: Might want to consider forcing the args into a tuple as you will get errors if it already one in this case, ie:

```
In [16]: fmt='hi %s'
In [17]: args='joe'
In [18]: fmt % args
Out[18]: 'hi joe'

In [20]: args='joe', 'smith'
In [21]: fmt % args
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-70797cd754bd> in <module>()
----> 1 fmt % args

TypeError: not all arguments converted during string formatting

In [22]: fmt % (args, )
Out[22]: "hi ('joe', 'smith')"
```

Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_disk.py]: Fixed, here and (I think) everywhere else I had the same idiom. Thank you!

**Comment by noahl** [comment]
@cwilkes I pushed a new version that I think addresses your comment. Please take a look.

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py [hunk 2]
```diff
@@ -98,132 +74,86 @@ def _Create(self):
...
-    # Just create a local directory at the specified path, don't mount
-    # anything.
-    self.RemoteCommand('sudo mkdir -p {0} && sudo chown -R $USER:$USER {0}'
-                       .format(disk_spec.mount_point))
-    self.scratch_disks.append(digitalocean_disk.DigitalOceanDisk(disk_spec))
+
+    if disk_spec.disk_type == disk.LOCAL:
+      if self.scratch_disks and self.scratch_disks[0].disk_type == disk.LOCAL:
+        # We have a disk object already, don't add more.
+        raise errors.Error('DigitalOcean does not support multiple disks.')
```
Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: does not support multiple local disk?

File: perfkitbenchmarker/providers/digitalocean/util.py [hunk 3]
```diff
@@ -15,98 +15,89 @@
...
-    args: key/value dict, used for JSON input
-    suppress_warning: if true, failures aren't interesting
   Returns:
-    (stdout, ret) tuple, where stdout is a JSON string and
-    ret is zero for success, the command exit code (1-255),
-    or the HTTP status code (300 and up).
-  """
-
-  cmd = GetCurlCommand(action, url, args)
+    A tuple of
```
Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/util.py]: Add Returns: before line 41

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py (hunk 2 above)
Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: No, all droplets just have one local SSD (as far as I know).

Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: Clarify the error message.

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py [hunk 4]
```diff
@@ -98,132 +74,86 @@ def _Create(self):
...
+
+      if disk_spec.num_striped_disks != 1:
+        raise ValueError('num_striped_disks=%s, but DigitalOcean VMs can only '
+                         'have one local disk.' % disk_spec.num_striped_disks)
+      # Since the local disk is already mounted, just create a local
+      # directory at the specified path. Calling
+      # self._CreateScratchDiskFromDisks would try to mount the
+      # volume, so don't do that.
```
Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: Add a comment explain why we want to mock boot disk as data disk (run on local ssds)

File: perfkitbenchmarker/providers/digitalocean/util.py [hunk 5]
```diff
@@ -15,98 +15,89 @@
 
 import json
 import logging
-import os
 
+from perfkitbenchmarker import errors
 from perfkitbenchmarker import flags
 from perfkitbenchmarker import vm_util
 
 DIGITALOCEAN_API = 'https://api.digitalocean.com/v2/'
```
Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/util.py]: Still needed>??

File: perfkitbenchmarker/providers/digitalocean/util.py [hunk 6]
```diff
@@ -15,98 +15,89 @@
...
+    errors.VmUtil.CalledProcessError if doctl fails.
+  """
 
-  return body, status
+  stdout, _, retcode = vm_util.IssueCommand(
+      ['doctl'] +
+      [str(arg) for arg in arg_list] +
+      ['--output=json'])
+
+  # In case of error, doctl sometimes prints "null" before printing
```
Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/util.py]: Add a todo to improve parsing

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py [hunk 7]
```diff
@@ -98,132 +74,86 @@ def _Create(self):
...
-    # Wait for completion. Actions are global objects, so the action
-    # status should still be retrievable after the VM got destroyed.
-    # We don't care about the result, let the _Exists check decide if we
-    # need to try again.
-    self._GetActionResult(destroy['id'])
+    response, retcode = util.DoctlAndParse(
+        ['compute', 'droplet', 'delete', self.droplet_id])
+    # The command doesn't return the HTTP status code, and the error
+    # format is very difficult to parse, so we string search.
+    if retcode and '404' in response['errors'][0]['detail']:
```
Comment (yuyantingzero) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: Add a TODO to parse the message?

**Comment by yuyantingzero** [comment]
LGTM. Mostly small comments on readability, feel free to merge after addressing

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py (hunk 2 above)
Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: Done, thanks.

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py (hunk 4 above)
Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: Added.

File: perfkitbenchmarker/providers/digitalocean/util.py (hunk 5 above)
Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/util.py]: No, removed. Thanks!

File: perfkitbenchmarker/providers/digitalocean/util.py (hunk 6 above)
Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/util.py]: Added.

File: perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py (hunk 7 above)
Comment (noahl) [inline perfkitbenchmarker/providers/digitalocean/digitalocean_virtual_machine.py]: Added.

**Comment by noahl** [comment]
Thanks for the review! I updated in response to your comments, and now I am merging.

**Pull request merged**